### PR TITLE
ROX-25572: prefetch dependencies for operator-bundle on Konflux

### PR DESCRIPTION
**Pull request by tommartensen**
### Description

Prefetch the Python dependencies required to run `patch-csv.py` bundle helpers script. 
This will fill the CSV template with information about related images, ... 

We can switch to the Python image as builder, because we don't use Go anymore to construct, patch and test the CSV.

### User-facing documentation

- [x] CHANGELOG update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request)  is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [x] existing tests suffice

#### How I validated my change

* Create a intermediate Docker image that has all RPM dependencies prefetched, like

```
FROM registry.access.redhat.com/ubi9:latest AS ubi-repo-donor

FROM brew.registry.redhat.io/rh-osbs/openshift-golang-builder:rhel_9_1.22 AS builder-runner

# For some reason, openshift-golang-builder 9 comes without any RPM repos in /etc/yum.repos.d/
# We, however, need to install some packages and so we need to configure RPM repos. The ones for UBI are sufficient.
COPY --from=ubi-repo-donor /etc/yum.repos.d/ubi.repo /etc/yum.repos.d/ubi.repo
RUN dnf -y upgrade --nobest && dnf -y install --nodocs --noplugins jq python3-pip

COPY ./operator/bundle_helpers/requirements.txt /tmp/requirements.txt
RUN pip3 install --no-cache-dir --only-binary=:all: -r /tmp/requirements.txt && rm /tmp/requirements.txt
```

* Use this as first stage in the `konflux.bundle.Dockerfile`
* Set `hermetic: true` in bundle build pipeline
* (remove waiting for other images, just to speed things up)

Validation build: https://console.redhat.com/application-pipeline/workspaces/rh-acs/applications/acs/pipelineruns/operator-bundle-on-push-hcgr2 (Commit: https://github.com/stackrox/stackrox/pull/13209/commits/f6c5ac0c9f6386b1541686ac860f9fc801928066) 

Then reverted those changes and ran a full validation pipeline: https://console.redhat.com/application-pipeline/workspaces/rh-acs/applications/acs/pipelineruns/operator-bundle-on-push-cxsf8 

```
[1/2] STEP 4/44: RUN . /cachi2/cachi2.env &&     pip3 install --no-cache-dir --only-binary=:all: -r /tmp/requirements.txt
Looking in links: /cachi2/output/deps/pip
Ignoring atomicwrites: markers 'python_version >= "3.6" and python_version < "4.0" and sys_platform == "win32"' don't match your environment
Ignoring colorama: markers 'python_version >= "3.6" and python_version < "4.0" and sys_platform == "win32"' don't match your environment
Ignoring importlib-metadata: markers 'python_version >= "3.6" and python_version < "3.8"' don't match your environment
Ignoring typing-extensions: markers 'python_version >= "3.6" and python_version < "3.8"' don't match your environment
Ignoring zipp: markers 'python_version >= "3.6" and python_version < "3.8"' don't match your environment
Processing /cachi2/output/deps/pip/attrs-22.2.0-py3-none-any.whl (from -r /tmp/requirements.txt (line 3))
Processing /cachi2/output/deps/pip/iniconfig-1.1.1-py2.py3-none-any.whl (from -r /tmp/requirements.txt (line 12))
Processing /cachi2/output/deps/pip/packaging-21.3-py3-none-any.whl (from -r /tmp/requirements.txt (line 15))
Processing /cachi2/output/deps/pip/pluggy-1.0.0-py2.py3-none-any.whl (from -r /tmp/requirements.txt (line 18))
Processing /cachi2/output/deps/pip/py-1.11.0-py2.py3-none-any.whl (from -r /tmp/requirements.txt (line 21))
Processing /cachi2/output/deps/pip/pyparsing-3.0.7-py3-none-any.whl (from -r /tmp/requirements.txt (line 24))
Processing /cachi2/output/deps/pip/pytest-7.0.1-py3-none-any.whl (from -r /tmp/requirements.txt (line 27))
Processing /cachi2/output/deps/pip/PyYAML-6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (from -r /tmp/requirements.txt (line 30))
Processing /cachi2/output/deps/pip/tomli-1.2.3-py3-none-any.whl (from -r /tmp/requirements.txt (line 71))
Installing collected packages: iniconfig, tomli, pyyaml, pyparsing, py, pluggy, attrs, packaging, pytest
Successfully installed attrs-22.2.0 iniconfig-1.1.1 packaging-21.3 pluggy-1.0.0 py-1.11.0 pyparsing-3.0.7 pytest-7.0.1 pyyaml-6.0 tomli-1.2.3
```

We see that `prefetch-dependencies` fetched our required dependencies and artifacts were used during the build.

### CODE REVIEW

File: operator/bundle_helpers/pyproject.toml
Comment (msugakov) [inline operator/bundle_helpers/pyproject.toml:4]:
```suggestion
description = "StackRox Operator Bundle Helpers"
```

File: operator/bundle_helpers/README.md
Comment (msugakov) [inline operator/bundle_helpers/README.md:3]:
```suggestion
## Python Dependency Management
```

File: operator/bundle_helpers/README.md
Comment (msugakov) [inline operator/bundle_helpers/README.md:13]: It deserves a mention that for as long as we have Downstream builds (hopefully not for long but still), all dependencies _actually_ used by scripts must be provided OOTB in CPaaS `render_templates` context.
CPaaS does not respect `requirements.txt` or anything and we can only use things that image maintainers included.
PyYAML is one such thing provided there and we can safely use it.

I did some basic thing to dump the python's version and list packages available in CPaaS but it was a while ago and things have changed.

I mention this because when you write `to add a new dependency`, it should be noted that we can't simply add any dependencies (and select any versions) we want while CPaaS is there.

Could you please somehow transform this in the note next to this list item?

**Review by msugakov — COMMENTED** [review]
Looks good! Please enable hermetic and I'll gladly approve

File: operator/bundle_helpers/README.md
Comment (tommartensen) [reply operator/bundle_helpers/README.md:16]: > CPaaS does not respect requirements.txt or anything and we can only use things that image maintainers included.

It used to do that before https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/commit/743a32a48bf967522aba7db54f91bc918f5a1cd9#609875223a8e1e7c7bc513ef01085c5fdbe00401_69_64, so we could install more dependencies here, right? (this is a very theoretical point, as we plan to move away from CPaaS soon and are unlikely to bump or add dependencies.

File: operator/bundle_helpers/README.md
Comment (tommartensen) [reply operator/bundle_helpers/README.md:4]: https://github.com/stackrox/stackrox/pull/13209/commits/ab8529e746514f797378710cc56b0aee6e660bb7

done

File: operator/bundle_helpers/README.md
Comment (tommartensen) [reply operator/bundle_helpers/README.md]: https://github.com/stackrox/stackrox/pull/13209/commits/ab8529e746514f797378710cc56b0aee6e660bb7

done

File: operator/bundle_helpers/pyproject.toml
Comment (tommartensen) [reply operator/bundle_helpers/pyproject.toml]: https://github.com/stackrox/stackrox/pull/13209/commits/ab8529e746514f797378710cc56b0aee6e660bb7

done

File: .tekton/operator-bundle-build.yaml
Comment (tommartensen) [reply .tekton/operator-bundle-build.yaml]: Yes!

https://github.com/stackrox/stackrox/pull/13209/commits/ab8529e746514f797378710cc56b0aee6e660bb7

done

File: operator/bundle_helpers/README.md
Comment (msugakov) [reply operator/bundle_helpers/README.md:16]: Cool that you remember!

In fact, what I wrote isn't correct in a practical sense: we could install `requirements.txt`, download anything from internet and do whatever else in `render_templates` (as well as in `pre-build-script`). But doing so violates the spirit of CPaaS&Co as a compliant and trusted build system.

These `render_templates`/`pre-build-script` unlike the Dockerfile builds which run in OSBS are executed in CPaaS Jenkins and aren't network-restricted. I think it's kind of intentional because some products need to do funky stuff for builds (like we download blobs) but at the same time these `render_templates`/`pre-build-script` aren't as hardened as Dockerfile builds in OSBS which can create supply chain security problems if we go and do what we want.

Nobody outside the team asked us to abandon `requirements.txt`, rather it was my idea to do that since we know that supply chain attacks on package repos happen regularly.

File: operator/bundle_helpers/README.md
Comment (msugakov) [inline operator/bundle_helpers/README.md:19]:
```suggestion
For as long as we build downstream images with CPaaS, you should not add or use dependencies that aren't available OOTB in CPaaS in `render_templates` context.
```

File: operator/bundle_helpers/README.md
Comment (msugakov) [inline operator/bundle_helpers/README.md:8]:
```suggestion
We use Poetry to manage Python dependencies and for generating the complete `requirements.txt`.
```

**Review by msugakov — APPROVED** [review]
Please address the new comments before merging.

File: operator/bundle_helpers/README.md
Comment (tommartensen) [reply operator/bundle_helpers/README.md:16]: I feel like this is resolved by https://github.com/stackrox/stackrox/pull/13209/files#diff-cc4b2f05c18ee7b629ad0bd4633ff59a26ffc67ad8607e97e3155c92f4b848eaR19 and will therefore resolve this comment.

**Pull request merged**
